test/python/transpiler/test_hoare_opt.py

```diff
@@ -0,0 +1,321 @@
+
+"""Test the HoareOptimizer pass"""
+
+import unittest
+from numpy import pi
+from qiskit.transpiler.passes import HoareOptimizer
+from qiskit.converters import circuit_to_dag
+from qiskit import QuantumCircuit
+from qiskit.test import QiskitTestCase
+from qiskit.extensions.standard import XGate, RZGate, CSwapGate, SwapGate
+from qiskit.extensions.unitary import UnitaryGate
+from qiskit.dagcircuit import DAGNode
+from qiskit.quantum_info import Statevector
+
+
+class TestHoareOptimizer(QiskitTestCase):
+    """Test the HoareOptimizer pass"""
+
+    def test_phasegate_removal(self):
+        """ Should remove the phase on a classical state,
+            but not on a superposition state.
+        """
+        circuit = QuantumCircuit(3)
+        circuit.z(0)
+        circuit.h(1)
+        circuit.z(1)
+
+        expected = QuantumCircuit(3)
+        expected.h(1)
+        expected.z(1)
+
+        sv = Statevector.from_label('0' * circuit.num_qubits)
+        (sv @ circuit) == (sv @ expected)
+
+        pass_ = HoareOptimizer(size=0)
+        result = pass_.run(circuit_to_dag(circuit))
+
+        self.assertEqual(result, circuit_to_dag(expected))
+
+    def test_cswap_removal(self):
+        """ Should remove Fredkin gates because the optimizer
+            can deduce the targets are in the same state
+        """
+        circuit = QuantumCircuit(10)
+        # prep
+        circuit.x(0)
+        circuit.h(3)
+        circuit.h(4)
+        circuit.h(5)
+        circuit.h(6)
+        # find first non-zero bit of reg(3-6), store position in reg(1-2)
+        circuit.cx(3, 0)
+        circuit.ccx(0, 4, 1)
+        circuit.cx(1, 0)
+        circuit.ccx(0, 5, 2)
+        circuit.cx(2, 0)
+        circuit.ccx(0, 6, 1)
+        circuit.ccx(0, 6, 2)
+        circuit.ccx(1, 2, 0)
+        # shift circuit
+        circuit.cswap(1, 7, 8)
+        circuit.cswap(1, 8, 9)
+        circuit.cswap(1, 9, 3)
+        circuit.cswap(1, 3, 4)
+        circuit.cswap(1, 4, 5)
+        circuit.cswap(1, 5, 6)
+        circuit.cswap(2, 7, 9)
+        circuit.cswap(2, 8, 3)
+        circuit.cswap(2, 9, 4)
+        circuit.cswap(2, 3, 5)
+        circuit.cswap(2, 4, 6)
+
+        expected = QuantumCircuit(10)
+        # prep
+        expected.x(0)
+        expected.h(3)
+        expected.h(4)
+        expected.h(5)
+        expected.h(6)
+        # find first non-zero bit of reg(3-6), store position in reg(1-2)
+        expected.cx(3, 0)
+        expected.ccx(0, 4, 1)
+        expected.cx(1, 0)
+        expected.ccx(0, 5, 2)
+        expected.cx(2, 0)
+        expected.ccx(0, 6, 1)
+        expected.ccx(0, 6, 2)
+        expected.ccx(1, 2, 0)
+        # optimized shift circuit
+        expected.cswap(1, 3, 4)
+        expected.cswap(1, 4, 5)
+        expected.cswap(1, 5, 6)
+        expected.cswap(2, 3, 5)
+        expected.cswap(2, 4, 6)
+
+        sv = Statevector.from_label('0' * circuit.num_qubits)
+        (sv @ circuit) == (sv @ expected)
```

Comment:
sorry I mistyped this the first time (edited later). This should be changed everywhere to
```
self.assertEqual(sv @ circuit, sv @ expected)
```
so it actually asserts something. Right now it doesn't do anything.

Reply:
done